src/library/scanner/libraryscanner.h

```diff
@@ -83,14 +90,35 @@ class LibraryScanner : public QThread {
     void slotFinishScan();
 
     // ScannerTask signal handlers.
-    void taskDone(bool success);
-    void directoryHashedAndScanned(const QString& directoryPath,
+    void slotTaskDone(bool success);
+    void slotDirectoryHashedAndScanned(const QString& directoryPath,
                                    bool newDirectory, int hash);
-    void directoryUnchanged(const QString& directoryPath);
-    void trackExists(const QString& trackPath);
-    void addNewTrack(TrackPointer pTrack);
+    void slotDirectoryUnchanged(const QString& directoryPath);
+    void slotTrackExists(const QString& trackPath);
+    void slotAddNewTrack(TrackPointer pTrack);
 
   private:
+    enum ScannerState {
+        IDLE,
+        STARTING,
+        SCANNING,
+        CANCELING,
+        FINISHED
+    };
+
+    void cancelAndQuit();
+    void cancel();
+
+    // Allowed State transitions:
```

Comment:
I have tried to consider all possible transactions and catch the not allowed by assertions. 
Do I miss one?